_posts/2023-09-11-range-joins-in-duckdb.md

```diff
@@ -0,0 +1,646 @@
+CREATE TABLE probe1 AS
+	SELECT k, '2022-01-01'::TIMESTAMP + INTERVAL (customer_id) HOUR purchase_timestamp
+	FROM range(10000) cs(k);
+```
+
+The build tables are much larger and have approximately
+10/100/1000x the number of entries as the 15 element tables:
+
+```sql
+-- 10:1
+CREATE TABLE build10 AS
+	SELECT k, t, (RANDOM() * 1000)::DECIMAL(7,2) AS v
+	FROM range(10000) ks(k), 
+	     range('2022-01-01'::TIMESTAMP, '2023-01-01'::TIMESTAMP, INTERVAL 59 HOUR) ts(t);
+-- 100:1
+CREATE TABLE build100 AS
+	SELECT k, t, (RANDOM() * 1000)::DECIMAL(7,2) AS v
+	FROM range(10000) ks(k), 
+	     range('2022-01-01'::TIMESTAMP, '2023-01-01'::TIMESTAMP, INTERVAL 350 MINUTE) ts(t);
+-- 1000:1
+CREATE TABLE build1000 AS
+	SELECT k, t, (RANDOM() * 1000)::DECIMAL(7,2) AS v
+	FROM range(10000) ks(k), 
+	     range('2022-01-01'::TIMESTAMP, '2023-01-01'::TIMESTAMP, INTERVAL 35 MINUTE) ts(t);
+```
+
+The AsOf join queries are:
+
+```sql
+-- AsOf/IEJoin
+SELECT p.k, p.t, v
+FROM probe p ASOF JOIN build b
+  ON p.k = b.k
+ AND p.t >= b.t
+ORDER BY 1, 2
+-- Rank
+WITH win AS (
+SELECT p.k, p.t, v,
+    rank() OVER (PARTITION BY p.k, p.t ORDER BY b.t DESC)  AS r
+FROM probe p INNER JOIN build b
+  ON p.k = b.k
+ AND p.t >= b.t
+QUALIFY r = 1
+) 
+SELECT k, t, v
+FROM win
+ORDER BY 1, 2
+```
+
+The results are shown here:
+
+<img src="/images/asof-rank.png"
+     alt="Rank Benchmark Results"
+     width="760"
+/>
+
+(Median of 5 except for Rank/15/1000). 
+
+* For all ratios with 15 probes, AsOf is the most performant.
+* For small ratios with 15 probes, Rank beats IEJoin (both with windowing), but by 100:1 it is starting to explode. 
+* For single element probes, Rank is most effective, but even there, its edge over AsOf is only about 50% at scale.
+
+This shows that AsOf could be possibly be improved upon, but predicting where that happens would be tricky, 
+and getting it wrong would have enormous costs.
+
+## Future Work
+
+DuckDB can now execute AsOf joins for all inequality types with reasonable performance.
+In some cases, the performance gain is several orders of magnitude over the standard SQL versions –
+even with our fast inequality join operator.
+
+While the current AsOf operator is completely general, 
+there are a couple of planning optimisations that could be applied here.
+
+* When there are selective equality conditions, it is likely that a hash join with filtering against a materialised state table would be significantly faster. If we can detect this and suitable sentinel values are available, the planner could choose to use a hash join instead of the default AsOf implementation.
+* There are also use cases where the probe table is much smaller than the build table, along with equality conditions, and performing a hash join against the *probe* table could yield significant performance improvements.
+
+Nevertheless, remember that one of the advantages of SQL is that it is a declarative language:  
```

Comment:
Colon at the end of the line is triggering a line break?

Reply:
It's worse :). It's the _space at the end of the line_ that's triggering a `<br>`-level line break (i.e. a line break but not a new paragraph). See https://github.com/mattcone/markdown-guide/blob/master/_basic-syntax/line-breaks.md#line-break-best-practices